torch/_inductor/codegen/cpp_bmm_template.py

```diff
@@ -0,0 +1,249 @@
+# mypy: allow-untyped-defs
+import contextlib
+from typing import Any, Callable, Dict, List, Optional, Tuple
+from unittest.mock import patch
+
+import sympy
+
+from .. import ir, lowering as L
+from ..select_algorithm import PartialRender
+from ..virtualized import V
+from .cpp_gemm_template import (
+    CppGemmTemplate,
+    GEMM_TEMPLATE,
+    get_padded_n,
+    MICROKERNEL_DEF,
+)
+from .cpp_micro_gemm import LayoutType
+from .cpp_template_kernel import CppTemplateKernel
+from .cpp_utils import DTYPE_TO_CPP, GemmBlocking
+
+
+GEMM_SINGLE_THREAD_MM_STUB = r"""
+{{kernel.def_kernel(
+    inputs={"X": X, "W": W},
+    outputs={"Y": Y},
+    aliases=aliases,
+    function_name="single_thread_mm",
+    symbols=[b_index],
+    placeholder="<SINGLE_THREAD_DEF_MM_FOR_BMM>")}}"""
+
+GEMM_THREADED_MM_STUB = r"""
+{{kernel.def_kernel(
+    inputs={"X": X, "W": W},
+    outputs={"Y": Y},
+    aliases=aliases,
+    function_name="threaded_mm",
+    symbols=[b_index],
+    placeholder="<THREADED_MM_DEF_FOR_BMM>")}}"""
+
+BMM_WRAPPER = r"""
+extern "C"
+{{kernel.def_kernel(inputs={"X": BX, "W": BW}, outputs={"Y": BY}, aliases=aliases)}}
+{
+    const int64_t B = {{kernel.size(BY_2d, 0)}};
+    {%- if num_threads > 1 %}
+    constexpr int64_t num_threads = {{num_threads}};
+    int64_t B_single_thread_block = (B / num_threads) * num_threads;
+
+    #pragma omp parallel for num_threads({{num_threads}})
+    {%- else %}
+    int64_t B_single_thread_block = B;
+    {%- endif %}
+    for (int64_t b_start = 0; b_start < B_single_thread_block; ++b_start) {
+        {{template.get_gemm_function_call(
+            kernel,
+            "single_thread_mm",
+            "<SINGLE_THREAD_CALL_FOR_BMM>",
+            b_index="b_start",
```

Comment:
May I know why this `b_index` is needed here? Can we slice tensors here directly?

Reply:
Earlier version of the code sliced the tensors directly. However, this created a problem for two cases with epilogue nodes: 
1. In some cases epilogue inputs were not indexed the same way as the tensors (for example, a 2D input of [B*M*N, 3]) with no clear indicator of what part should be sliced here. 
2. For binary ops, the kernel has recorded the name of the epilogue node input buffer, but the actual buffer object was not available in the kernel to be sliced.

For both of these cases, passing an index variable and slicing the tensors in lines 201-202 fixes any problem with epilogue indexing.